File: 1508-longest-happy-prefix/longest-happy-prefix.cpp
class Solution {
public:
void lps(vector<int>&ans,string s){
    int m=s.size();
int pre=0;
int suf=1;
while(suf<m){
    if(s[pre]==s[suf]){
        ans[suf]=pre+1;
        pre++;
        suf++;
    }
    else{
        if(pre==0){
            ans[suf]=0;
            suf++;
        }
        else{
            pre=ans[pre-1];//"Agar pura match fail ho gaya, toh uske andar ka best partial match try karo" 
        }
    }
}

}
    string longestPrefix(string s) {
        int n=s.size();
        vector<int>ans(n,0);
        lps(ans,s);
        int len = ans[n-1];
         return s.substr(0, len);
    }
};

File: 1508-longest-happy-prefix/longest-happy-prefix.java
class Solution {
    public String longestPrefix(String s) {
        int n=s.length();
        int[] lps=new int[n];
        int pre=0;
        int suff=1;
        while(suff<n){
            if(s.charAt(pre)==s.charAt(suff)){
                pre++;
                lps[suff]=pre;
                suff++;
            }else if(pre==0){
                lps[suff]=0;
                suff++;
            }else{
                pre=lps[pre-1];
            }
        }
        return s.substring(0,lps[n-1]);
    }
}